### PR TITLE
Re-do sort order, fix scanner timeout, fix reaper.

### PR DESCRIPTION
**Pull request by athornton**

### CODE REVIEW

**Comment by frossie** [comment]
I'd like to avoid date parsing if possible.  It's too fragile. 
1. For the official releases in the pre-pulled (in-page) sections, it can be simpler than all this: take the list of officials, throw away the rcs (putting up the release candidate defeats the purpose of "latest official release"), use distutils.version in the standard library to sort
2. For the non-prepulled dropdowns: make separate dropdowns for (a) officials and release candidates (b) d and exps (c) weeklies - you can sort those without date parsing and it makes easier to find things as they are getting quite long anyway. You can use distutils.version to sort the rcs, and simple sorting (by stripping the exp/d/etc) for the d/exps and the w/s. That will solve your problem of your exps being at the bottom while also keeping the users away from them. 
3. Remove code marked "thoroughly obsolete" if that is the case or fix comment if otherwise
4. Send me another PR, there's some minor things I can have another pass at once the date sorting is gone. 

Ideally the pipeline sorting code would be somewhere unit testable that pipelines can also fiddle with, but let's forget that for this PR.

**Comment by athornton** [comment]
I think the idea of multiple dropdowns is not a good one.  That's a lot of screen real estate for a fairly rare case, and I'm not sure how it would work if a user was clicking among them and had images selected in more than one dropdown.

We should probably talk about this Monday.

I think we *do* want to prepull and display a release candidate build if it is newer than any of the official releases.  You gotta test rcs too and if it's not prepulled that's very painful.  I like the idea of never displaying or prepulling rc versions after the official one is released.

Experimentals can also have weekly tags, so there's still got to be some way of determining their ordering if we want to categorize them with dailies.  We build experimental weeklies of jlbleed every week.

**Comment by frossie** [comment]
Let's talk about it Monday then. 
As to the other point - an official release says "we produced official data products with this release and have tested it to within an inch of its life". A release candidate is by definition not that. Also, a release candidate is always a weekly anyway, so internal candidates have easy access to it.

**Comment by athornton** [comment]
So, after some consideration, semver (rather than distutils.version) seemed like the better idea to me.  Same plan, though: given a tag, extract a semantic version from it, that is comparable to other semantic versions.  We're already importing semver in the module so I didn't increase my dependency list.

I'm currently still doing year/week -> year/month/day for semver purposes.  If you don't want even that, we can (at the price of making the all-tags sort order less chronological and more categorical) pick a minor-version offset greater than 12 for the week and have non-overlapping versions for weekly and daily builds in a given year, or make the weekly builds have a major version offset and have them sorted entirely separately from weekly and daily.  All the releases are currently at the bottom of the list, but again we can change that by manipulating the semver major version.  Since the version isn't used outside the context of sorting the results, we don't actually care if the major version really matches the year of production or not (which it already doesn't for releases).

The major trick here is that I am storing a _reduced_results list within the scanner now, which is sorted by semver, but also includes a type field.  So getting all the weeklies (or any category) comes down to [ x for x in self._reduced_results if x['type'] == 'weekly' ], and they're already ordered.  This enabled quite a bit of simplification although it's not obvious in the diff.  All the sorting just boils down, now, to sort(key=lambda x: x['version'], reverse=True) because 'version' is a semver.VersionInfo object which knows how to compare.

I threw in the "use release rcs if they are newer than the last real release but otherwise don't show 'em," but because semver (with a little lexigraphic help) recognizes "rcX" as setting a prerelease version, the logic is very clean.  At least until we have r_21_0_0_rc1one11!eleven!  Hiding anything with a prerelease tag is two lines of code less.

File: rubin_jupyter_utils/hub/scanrepo/scanrepo.py [hunk 1]
```diff
@@ -143,75 +141,188 @@ def _read_cachefile(self):
...
+        #
         ld = tag  # Default description is just the tag name
         components = None
-        if tag.find("_") != -1:
+        ttype = None
+        major = 0
+        minor = 0
+        patch = 0
+        rest = None
+        if tag.find("_") != -1:  # New-style tag
```
Comment (frossie) [inline rubin_jupyter_utils/hub/scanrepo/scanrepo.py]: There must be someway less c-like to write this for clarity - 
`if ‘_’ in tag:`
or something?

**Review by frossie — CHANGES_REQUESTED** [review]
Left a comment. Otherwise you can merge for now but there's some stuff we can discuss further at standup.

File: rubin_jupyter_utils/hub/scanrepo/scanrepo.py (hunk 1 above)
Comment (athornton) [reply rubin_jupyter_utils/hub/scanrepo/scanrepo.py]: That is indeed the less c-like way to write it.  Changed.

**Pull request merged**